sstables/sstables.cc

```diff
@@ -2776,12 +2776,21 @@ future<bool> sstable::has_partition_key(const utils::hashed_key& hk, const dht::
     std::exception_ptr ex;
     auto sem = reader_concurrency_semaphore(reader_concurrency_semaphore::no_limits{}, "sstables::has_partition_key()",
             reader_concurrency_semaphore::register_metrics::no);
+    bool reader_closed = true;  // Set to true while index_reader isn't created.
     try {
         auto lh_index_ptr = std::make_unique<sstables::index_reader>(s, sem.make_tracking_only_permit(_schema, s->get_filename(), db::no_timeout, {}));
-        present = co_await lh_index_ptr->advance_lower_and_check_if_present(dk);
+        reader_closed = false;
+        try {
+            present = co_await lh_index_ptr->advance_lower_and_check_if_present(dk);
+        } catch (...) {
+            ex = std::current_exception();
+        }
+        reader_closed = true;
+        co_await lh_index_ptr->close();
     } catch (...) {
         ex = std::current_exception();
     }
+    assert(reader_closed);  // Reader should be closed before the semaphore is stopped.
```

Comment:
This assertion looks quite silly right now, but actually at first I tried to use `.finally` and had it triggered. So I would leave it as is just in case someone's tempted to modify the code ;)